compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs

```diff
@@ -365,14 +366,54 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
                 }
 
                 self.infcx.probe(|_| {
-                    if let Ok(_substs) = self.match_impl(impl_def_id, impl_trait_ref, obligation) {
+                    if let Ok(_) = self.match_impl(impl_def_id, impl_trait_ref, obligation) {
+                        if self.reject_fn_ptr_impls(impl_def_id, obligation) {
```

Comment:
r=me on the general approach if perf is acceptable. Only need an FCP :sweat_smile: it kinda sucks that this kinda requires approval from all of types, lang, and libs :sweat_smile: 

I would prefer us to make it clear that this a hack only for perf and diagnostics, please add a comment here and rename the function to `fast_reject_x` :grin:

Reply:
I had it called `fast_..` first, but considered that invoking `evaluate_obligation` isn't really fast 😆